### PR TITLE
abstract and composite queues

### PR DESCRIPTION
**Pull request by y0s3ph91**

### CODE REVIEW

File: src/Task/CompositeQueue.php [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+
+namespace SunValley\MPW\Task;
+
+
+class CompositeQueue
+{
+    /** @var TaskQueueInterface[] */
+    protected $taskQueues = [];
+
+    public function setTaskQueues(TaskQueueInterface $queues)
```
Comment (bmlkc) [inline src/Task/CompositeQueue.php]: method name => addTaskQueue

Comment (bmlkc) [reply src/Task/CompositeQueue.php]: method body => addTaskQueue(TaskQueueInterface $queue, int $priority = 0)

Comment (bmlkc) [reply src/Task/CompositeQueue.php]: you also need to implement the other three methods here

Comment (y0s3ph91) [reply src/Task/CompositeQueue.php]: PR updated, take a look when you have some free time, thank you

File: src/Task/CompositeQueueInterface.php [hunk 2]
```diff
@@ -0,0 +1,44 @@
+<?php
+
+namespace SunValley\MPW\Task;
+
+
+interface CompositeQueueInterface
```
Comment (bmlkc) [inline src/Task/CompositeQueueInterface.php]: extends TaskQeueuInterface

File: src/Task/PrioritizedTaskQueueInterface.php [hunk 3]
```diff
@@ -0,0 +1,9 @@
+<?php
```
Comment (bmlkc) [inline src/Task/PrioritizedTaskQueueInterface.php]: this class should be removed (not adding anything)

File: src/Task/TaskQueueInterface.php [hunk 4]
```diff
@@ -39,8 +39,8 @@ function put(TaskInterface $task, int $priority = null);
     function count(): int;
 
     /**
-     * Provisional timeout for poll to give up and throw an exception 
-     * 
+     * Provisional timeout for poll to give up and throw an exception
+     *
      * @param int $timeout
      *
      * @return mixed
```
Comment (bmlkc) [inline src/Task/TaskQueueInterface.php]: returns void

File: src/Worker/PoolWorker.php [hunk 5]
```diff
@@ -4,26 +4,31 @@
 namespace SunValley\MPW\Worker;
 
 
+use SunValley\MPW\Task\CompositeQueueInterface;
```
Comment (bmlkc) [inline src/Worker/PoolWorker.php]: this is wrong

Comment (bmlkc) [reply src/Worker/PoolWorker.php]: composite queue should be an instance of TaskQueueInterface. the changes here is completely wrong. you should not even change this file

File: src/Worker/PoolWorkerInterface.php [hunk 6]
```diff
@@ -3,7 +3,7 @@
```
Comment (bmlkc) [inline src/Worker/PoolWorkerInterface.php]: wrong implementation

File: src/Task/AbstractCompositeQueue.php [hunk 7]
```diff
@@ -0,0 +1,33 @@
+<?php
+
+namespace SunValley\MPW\Task;
+
+
+abstract class AbstractCompositeQueue implements CompositeQueueInterface, PrioritizedTaskQueueInterface
```
Comment (bmlkc) [inline src/Task/AbstractCompositeQueue.php]: CompositeQueue is a concrete implementation of TaskQueueInterface

Comment (bmlkc) [reply src/Task/AbstractCompositeQueue.php]: cannot be abstract

File: src/Task/CompositeQueue.php [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+    }
+
+    /**
+     * Retrieves and removes the first task of this queue.
+     *
+     * @return TaskInterface
+     */
+    function poll(): TaskInterface
+    {
+        return array_shift($this->taskQueue);
```
Comment (bmlkc) [inline src/Task/CompositeQueue.php]: wrong, you need to return a task not taskQueue

File: src/Task/CompositeQueue.php [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+     * @param TaskInterface $task
+     * @param int|null      $priority Descending order, i.e. bigger numbers have precedence.  If tasks is a prioritized
+     *                                task and no priority is given, "default" priority from the task is used. Tasks
+     *                                with no priority definitions is registered as priority=0.
+     *
+     * @return void
+     */
+    function put(TaskInterface $task, int $priority = null)
+    {
+        // TODO: Implement put() method.
```
Comment (bmlkc) [inline src/Task/CompositeQueue.php]: required

Comment (bmlkc) [reply src/Task/CompositeQueue.php]: as putting a task to two task queues is not really a wanted behavior, you just have to throw an exception here stating that "composite queues cannot be used for submitting tasks".

**Pull request merged**